ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java

```diff
@@ -239,6 +240,42 @@ private KsqlSerdeFactory getValueSerdeFactory(final Sink sink) {
       return serdeFactories.create(format, sink.getProperties());
     }
 
+    /**
+     * Get the list of select expressions that are <i>not</i> for meta and key fields in the source
+     * schema.
+     *
+     * <p>Currently, the select expressions can include metadata and key fields, which are latter
+     * removed by {@link io.confluent.ksql.planner.plan.KsqlStructuredDataOutputNode}. When building
+     * the {@link SerdeOptions} its important the finally schema is used, i.e. the one with these
+     * meta and key fields removed.
+     *
+     * <p>This is weird functionality, but maintained for backwards compatibility for now.
+     *
+     * @return the list of field names in the sink that are not meta or key fields.
+     */
+    private List<String> getNoneMetaOrKeySelectAliases() {
+      final SourceSchemas sourceSchemas = analysis.getFromSourceSchemas();
+      final List<Expression> selects = analysis.getSelectExpressions();
+
+      final List<String> columnNames = new ArrayList<>(analysis.getSelectExpressionAlias());
+
+      for (int idx = selects.size() - 1; idx >= 0; --idx) {
+        final Expression select = selects.get(idx);
+
+        if (!(select instanceof DereferenceExpression)
+            && !(select instanceof QualifiedNameReference)) {
+          continue;
+        }
+
+        if (!sourceSchemas.matchesNonValueField(select.toString())) {
```

Comment:
we're calling `matchesNonValueField` on every select, and `matchesNonValueField` creates a new set of nonValue fields every time. why not just construct a Map<LogicalSchema, Set<String>> nonValueFIelds when we create `SourceSchemas`?

Reply:
Premature optimization is the root of _ALL_ evil...

This isn't on the critical path. It's only executed when analyzing queries.